subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/health/gc/GarbageCollectorMonitoringStrategy.java

```diff
@@ -85,35 +96,37 @@ public double getThrashingThreshold() {
+    }
+
+    public boolean isAboveNonHeapUsageThreshold(int percent) {
+        return nonHeapUsageThreshold != -1 && percent >= nonHeapUsageThreshold;
+    }
+
+    public boolean isAboveGcRateThreshold(double gcEventsPerSec) {
+        return gcRateThreshold != -1 && gcEventsPerSec >= gcRateThreshold;
+    }
+
+    public boolean isAboveGcThrashingThreshold(double gcEventsPerSec) {
+        return thrashingThreshold != -1 && gcEventsPerSec >= thrashingThreshold;
+    }
+
     public static GarbageCollectorMonitoringStrategy determineGcStrategy() {
-        final List<String> garbageCollectors = CollectionUtils.collect(ManagementFactory.getGarbageCollectorMXBeans(), new Transformer<String, GarbageCollectorMXBean>() {
-            @Override
-            public String transform(GarbageCollectorMXBean garbageCollectorMXBean) {
-                return garbageCollectorMXBean.getName();
-            }
-        });
-        GarbageCollectorMonitoringStrategy gcStrategy = CollectionUtils.findFirst(GarbageCollectorMonitoringStrategy.values(), new Spec<GarbageCollectorMonitoringStrategy>() {
-            @Override
-            public boolean isSatisfiedBy(GarbageCollectorMonitoringStrategy strategy) {
-                return garbageCollectors.contains(strategy.getGarbageCollectorName());
-            }
-        });
+        List<String> garbageCollectors = CollectionUtils.collect(ManagementFactory.getGarbageCollectorMXBeans(), MemoryManagerMXBean::getName);
+        GarbageCollectorMonitoringStrategy gcStrategy = CollectionUtils.findFirst(STRATEGIES, strategy -> garbageCollectors.contains(strategy.getGarbageCollectorName()));
 
         if (gcStrategy == null) {
             LOGGER.info("Unable to determine a garbage collection monitoring strategy for {}", Jvm.current());
             return GarbageCollectorMonitoringStrategy.UNKNOWN;
-        } else {
-            List<String> memoryPools = CollectionUtils.collect(ManagementFactory.getMemoryPoolMXBeans(), new Transformer<String, MemoryPoolMXBean>() {
-                @Override
-                public String transform(MemoryPoolMXBean memoryPoolMXBean) {
-                    return memoryPoolMXBean.getName();
-                }
-            });
-            if (!memoryPools.contains(gcStrategy.heapPoolName) || !memoryPools.contains(gcStrategy.nonHeapPoolName)) {
-                LOGGER.info("Unable to determine which memory pools to monitor for {}", Jvm.current());
-                return GarbageCollectorMonitoringStrategy.UNKNOWN;
-            }
-            return gcStrategy;
         }
+
+        List<String> memoryPools = CollectionUtils.collect(ManagementFactory.getMemoryPoolMXBeans(), MemoryPoolMXBean::getName);
+        if (!memoryPools.contains(gcStrategy.heapPoolName) || !memoryPools.contains(gcStrategy.nonHeapPoolName)) {
+            LOGGER.info("Unable to determine which memory pools to monitor for {}", Jvm.current());
```

Comment:
Yeah, this isn't really something the user would do something about, but it would be better to communicate the ramifications of the issue (i.e. that daemon heap monitoring will be disabled).  And perhaps suggest that they file an issue with details we gather (JDK, garbage collector name, strategy, pool names, etc).